File: stash_a8db530/unity_sdk/Runtime/QyntaraRuntimeGenerator.cs
/*
 * Qyntara AI Runtime Generator - Unity C# Wrapper
 * Enables real-time 3D generation in Unity using ONNX Runtime
 * 
 * v6.0 Prototype - Runtime SDK
 */

using UnityEngine;
using Unity.Barracuda;
using System.Collections;
using System.IO;

namespace Qyntara.Runtime
{
    /// <summary>
    /// Real-time 3D mesh generator for Unity using LRM (Large Reconstruction Model)
    /// </summary>
    public class QyntaraRuntimeGenerator : MonoBehaviour
    {
        [Header("Model Settings")]
        [Tooltip("Path to ONNX model file")]
        public string modelPath = "Assets/Qyntara/Models/lrm_model.onnx";
        
        [Tooltip("Input image resolution")]
        public int inputResolution = 512;
        
        [Header("Generation Settings")]
        [Tooltip("Marching cubes resolution (higher = more detail, slower)")]
        [Range(64, 512)]
        public int marchingCubesResolution = 256;
        
        [Tooltip("Auto-generate mesh on start")]
        public bool generateOnStart = false;
        
        [Header("Output")]
        [Tooltip("Generated mesh will be assigned here")]
        public MeshFilter outputMeshFilter;
        
        private Model runtimeModel;
        private IWorker worker;
        private bool isGenerating = false;
        
        void Start()
        {
            LoadModel();
            
            if (generateOnStart && outputMeshFilter != null)
            {
                // Generate from default texture
                Texture2D defaultTexture = new Texture2D(inputResolution, inputResolution);
                StartCoroutine(GenerateMeshAsync(defaultTexture));
            }
        }
        
        /// <summary>
        /// Load ONNX model for inference
        /// </summary>
        void LoadModel()
        {
            try
            {
                Debug.Log($"[Qyntara] Loading model from {modelPath}");
                
                // Load ONNX model using Barracuda
                runtimeModel = ModelLoader.Load(modelPath);
                worker = WorkerFactory.CreateWorker(WorkerFactory.Type.ComputePrecompiled, runtimeModel);
                
                Debug.Log("[Qyntara] Model loaded successfully!");
            }
            catch (System.Exception e)
            {
                Debug.LogError($"[Qyntara] Failed to load model: {e.Message}");
            }
        }
        
        /// <summary>
        /// Generate 3D mesh from texture asynchronously
        /// </summary>
        /// <param name="inputTexture">Input image texture</param>
        /// <returns>Coroutine</returns>
        public IEnumerator GenerateMeshAsync(Texture2D inputTexture)
        {
            if (isGenerating)
            {
                Debug.LogWarning("[Qyntara] Generation already in progress");
                yield break;
            }
            
            if (worker == null)
            {
                Debug.LogError("[Qyntara] Model not loaded");
                yield break;
            }
            
            isGenerating = true;
            float startTime = Time.realtimeSinceStartup;
            
            Debug.Log("[Qyntara] Starting mesh generation...");
            
            // Preprocess image
            Tensor inputTensor = PreprocessImage(inputTexture);
            
            // Run inference
            worker.Execute(inputTensor);
            Tensor outputTensor = worker.PeekOutput();
            
            // Wait for GPU to finish
            yield return new WaitForEndOfFrame();
            
            // Extract mesh from output
            Mesh generatedMesh = ExtractMesh(outputTensor);
            
            // Assign to output
            if (outputMeshFilter != null && generatedMesh != null)
            {
                outputMeshFilter.mesh = generatedMesh;
                Debug.Log($"[Qyntara] Mesh generated in {(Time.realtimeSinceStartup - startTime):F2}s");
                Debug.Log($"[Qyntara] Vertices: {generatedMesh.vertexCount}, Triangles: {generatedMesh.triangles.Length / 3}");
            }
            
            // Cleanup
            inputTensor.Dispose();
            outputTensor.Dispose();
            
            isGenerating = false;
        }
        
        /// <summary>
        /// Preprocess input image for model
        /// </summary>
        Tensor PreprocessImage(Texture2D texture)
        {
            // Resize to input resolution
            Texture2D resized = ResizeTexture(texture, inputResolution, inputResolution);
            
            // Convert to tensor (normalize to [-1, 1])
            float[] pixels = new float[inputResolution * inputResolution * 3];
            Color[] colors = resized.GetPixels();
            
            for (int i = 0; i < colors.Length; i++)
            {
                pixels[i * 3 + 0] = colors[i].r * 2f - 1f;  // R
                pixels[i * 3 + 1] = colors[i].g * 2f - 1f;  // G
                pixels[i * 3 + 2] = colors[i].b * 2f - 1f;  // B
            }
            
            return new Tensor(1, inputResolution, inputResolution, 3, pixels);
        }
        
        /// <summary>
        /// Extract mesh from model output tensor
        /// </summary>
        Mesh ExtractMesh(Tensor outputTensor)
        {
            // NOTE: This is a PROTOTYPE
            // Real implementation would decode the tensor into mesh data
            // For now, create a simple cube as placeholder
            
            Mesh mesh = new Mesh();
            
            // Simple cube vertices
            Vector3[] vertices = new Vector3[]
            {
                new Vector3(-0.5f, -0.5f, -0.5f),
                new Vector3(0.5f, -0.5f, -0.5f),
                new Vector3(0.5f, 0.5f, -0.5f),
                new Vector3(-0.5f, 0.5f, -0.5f),
                new Vector3(-0.5f, -0.5f, 0.5f),
                new Vector3(0.5f, -0.5f, 0.5f),
                new Vector3(0.5f, 0.5f, 0.5f),
                new Vector3(-0.5f, 0.5f, 0.5f)
            };
            
            int[] triangles = new int[]
            {
                0, 2, 1, 0, 3, 2,
                4, 5, 6, 4, 6, 7,
                0, 1, 5, 0, 5, 4,
                1, 2, 6, 1, 6, 5,
                2, 3, 7, 2, 7, 6,
                3, 0, 4, 3, 4, 7
            };
            
            mesh.vertices = vertices;
            mesh.triangles = triangles;
            mesh.RecalculateNormals();
            mesh.RecalculateBounds();
            
            return mesh;
        }
        
        /// <summary>
        /// Resize texture to target resolution
        /// </summary>
        Texture2D ResizeTexture(Texture2D source, int targetWidth, int targetHeight)
        {
            RenderTexture rt = RenderTexture.GetTemporary(targetWidth, targetHeight);
            Graphics.Blit(source, rt);
            
            RenderTexture previous = RenderTexture.active;
            RenderTexture.active = rt;
            
            Texture2D result = new Texture2D(targetWidth, targetHeight);
            result.ReadPixels(new Rect(0, 0, targetWidth, targetHeight), 0, 0);
            result.Apply();
            
            RenderTexture.active = previous;
            RenderTexture.ReleaseTemporary(rt);
            
            return result;
        }
        
        void OnDestroy()
        {
            worker?.Dispose();
        }
    }
}

File: stash_a8db530/threejs_sdk/src/QyntaraRuntimeGenerator.ts
/**
 * Qyntara AI Runtime Generator - Three.js WebGL Wrapper
 * Enables real-time 3D generation in web browsers using ONNX.js
 * 
 * v6.0 Prototype - Runtime SDK
 */

import * as THREE from 'three';
import * as ort from 'onnxruntime-web';

export interface GenerationOptions {
    modelPath: string;
    inputResolution: number;
    marchingCubesResolution: number;
}

export class QyntaraRuntimeGenerator {
    private session: ort.InferenceSession | null = null;
    private isGenerating: boolean = false;
    private modelLoaded: boolean = false;

    constructor(private options: GenerationOptions) {
        this.options = {
            modelPath: options.modelPath || '/models/lrm_model.onnx',
            inputResolution: options.inputResolution || 512,
            marchingCubesResolution: options.marchingCubesResolution || 256
        };
    }

    /**
     * Load ONNX model for inference
     */
    async loadModel(): Promise<void> {
        try {
            console.log('[Qyntara] Loading model from', this.options.modelPath);

            // Configure ONNX Runtime for WebGL
            ort.env.wasm.wasmPaths = '/onnx-wasm/';

            // Load model
            this.session = await ort.InferenceSession.create(this.options.modelPath, {
                executionProviders: ['webgl', 'wasm']
            });

            this.modelLoaded = true;
            console.log('[Qyntara] Model loaded successfully!');
        } catch (error) {
            console.error('[Qyntara] Failed to load model:', error);
            throw error;
        }
    }

    /**
     * Generate 3D mesh from image
     */
    async generateMesh(imageElement: HTMLImageElement): Promise<THREE.Mesh> {
        if (this.isGenerating) {
            throw new Error('Generation already in progress');
        }

        if (!this.modelLoaded || !this.session) {
            throw new Error('Model not loaded');
        }

        this.isGenerating = true;
        const startTime = performance.now();

        try {
            console.log('[Qyntara] Starting mesh generation...');

            // Preprocess image
            const inputTensor = this.preprocessImage(imageElement);

            // Run inference
            const feeds = { image: inputTensor };
            const results = await this.session.run(feeds);
            const outputTensor = results.mesh_features;

            // Extract mesh from output
            const mesh = this.extractMesh(outputTensor);

            const elapsedTime = (performance.now() - startTime) / 1000;
            console.log(`[Qyntara] Mesh generated in ${elapsedTime.toFixed(2)}s`);

            return mesh;
        } catch (error) {
            console.error('[Qyntara] Generation failed:', error);
            throw error;
        } finally {
            this.isGenerating = false;
        }
    }

    /**
     * Preprocess image for model input
     */
    private preprocessImage(imageElement: HTMLImageElement): ort.Tensor {
        const canvas = document.createElement('canvas');
        const ctx = canvas.getContext('2d')!;

        canvas.width = this.options.inputResolution;
        canvas.height = this.options.inputResolution;

        // Draw and resize image
        ctx.drawImage(imageElement, 0, 0, canvas.width, canvas.height);

        // Get pixel data
        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
        const pixels = imageData.data;

        // Convert to float32 array and normalize to [-1, 1]
        const inputSize = this.options.inputResolution * this.options.inputResolution * 3;
        const inputData = new Float32Array(inputSize);

        for (let i = 0; i < pixels.length / 4; i++) {
            inputData[i * 3 + 0] = (pixels[i * 4 + 0] / 255.0) * 2 - 1; // R
            inputData[i * 3 + 1] = (pixels[i * 4 + 1] / 255.0) * 2 - 1; // G
            inputData[i * 3 + 2] = (pixels[i * 4 + 2] / 255.0) * 2 - 1; // B
        }

        // Create tensor (NCHW format)
        return new ort.Tensor('float32', inputData, [
            1,
            3,
            this.options.inputResolution,
            this.options.inputResolution
        ]);
    }

    /**
     * Extract Three.js mesh from model output
     */
    private extractMesh(outputTensor: ort.Tensor): THREE.Mesh {
        // NOTE: This is a PROTOTYPE
        // Real implementation would decode the tensor into mesh data
        // For now, create a simple cube as placeholder

        const geometry = new THREE.BoxGeometry(1, 1, 1);
        const material = new THREE.MeshStandardMaterial({
            color: 0x00ff00,
            metalness: 0.5,
            roughness: 0.5
        });

        const mesh = new THREE.Mesh(geometry, material);

        console.log('[Qyntara] Created placeholder cube mesh (PROTOTYPE)');
        return mesh;
    }

    /**
     * Check if generation is in progress
     */
    isGeneratingMesh(): boolean {
        return this.isGenerating;
    }

    /**
     * Dispose resources
     */
    dispose(): void {
        if (this.session) {
            this.session = null;
            this.modelLoaded = false;
        }
    }
}

/**
 * Example usage:
 * 
 * import { QyntaraRuntimeGenerator } from './QyntaraRuntimeGenerator';
 * 
 * const generator = new QyntaraRuntimeGenerator({
 *     modelPath: '/models/lrm_model.onnx',
 *     inputResolution: 512,
 *     marchingCubesResolution: 256
 * });
 * 
 * await generator.loadModel();
 * 
 * const image = document.getElementById('input-image') as HTMLImageElement;
 * const mesh = await generator.generateMesh(image);
 * 
 * scene.add(mesh);
 */
